presto-main/src/main/java/com/facebook/presto/operator/GroupIdOperator.java

```diff
@@ -0,0 +1,239 @@
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        checkState(!finishing, "Operator is already finishing");
+        requireNonNull(page, "page is null");
+
+        groupIdRowGenerator.addNewPage(page);
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        if (!groupIdRowGenerator.hasBufferedRows()) {
+            return null;
+        }
+
+        return groupIdRowGenerator.nextPage();
+    }
+
+    private static class GroupIdRowGenerator
+    {
+        private final BitSet[] groupingSetNullChannels;
+        private final PageBuilder pageBuilder;
+        private final int numGroupingSets;
+        private Page currentPage = null;
+        private int currentPosition = 0;
+
+        private GroupIdRowGenerator(List<Type> types, BitSet[] groupingSetNullChannels)
+        {
+            this.pageBuilder = new PageBuilder(types);
+            this.groupingSetNullChannels = groupingSetNullChannels;
+            this.numGroupingSets = groupingSetNullChannels.length;
+        }
+
+        private void addNewPage(Page page)
+        {
+            checkState(currentPage == null, "currentPage must be null to add a new page");
+            currentPage = page;
+            currentPosition = 0;
+        }
+
+        private void process()
+        {
+            while (currentPosition < currentPage.getPositionCount() && !pageBuilder.isFull()) {
+                // generate 'n' rows for every input row, where n is the number of grouping sets
+                for (int groupingSet = 0; groupingSet < numGroupingSets; groupingSet++) {
+                    pageBuilder.declarePosition();
+
+                    // write group ID
+                    BIGINT.writeLong(pageBuilder.getBlockBuilder(0), groupingSet);
+
+                    for (int channel = 0; channel < currentPage.getChannelCount(); channel++) {
+                        // if column is to be ignored or is already null, set output column to be null
+                        if (groupingSetNullChannels[groupingSet].get(channel) || currentPage.getBlock(channel).isNull(currentPosition)) {
+                            pageBuilder.getBlockBuilder(channel + 1).appendNull();
+                            continue;
+                        }
+
+                        // copy everything else as-is
+                        currentPage.getBlock(channel).writePositionTo(currentPosition, pageBuilder.getBlockBuilder(channel + 1));
+                        pageBuilder.getBlockBuilder(channel + 1).closeEntry();
+                    }
+                }
+
+                currentPosition++;
+            }
+
+            // if we finished processing this page
+            if (currentPosition == currentPage.getPositionCount()) {
+                currentPage = null;
+            }
+        }
+
+        public boolean hasBufferedRows()
+        {
+            return !(currentPage == null && pageBuilder.isEmpty());
```

Comment:
pageBuilder can be empty when getOutput is called twice and no additional input is given (i.e. it just finished processing a single input page)